crates/turborepo-lib/src/daemon/server.rs

```diff
@@ -0,0 +1,313 @@
+        let watcher_fut = watcher.watch(repo_root.to_path_buf(), stop.token());
+
+        let timer = self.timeout_time.clone();
+        let timeout_fut = timer.wait();
+
+        // if shutdown is available, then listen. otherwise just wait forever
+        let shutdown_rx = self.shutdown_rx.take();
+        let shutdown_fut = async move {
+            match shutdown_rx {
+                Some(rx) => {
+                    rx.await.ok();
+                }
+                None => {
+                    futures::pending!();
+                }
+            }
+        };
+
+        #[cfg(feature = "http")]
+        let server_fut = {
+            // set up grpc reflection
+            let efd = include_bytes!("file_descriptor_set.bin");
+            let reflection = tonic_reflection::server::Builder::configure()
+                .register_encoded_file_descriptor_set(efd)
+                .build()
+                .unwrap();
+
+            Server::builder()
+                .add_service(reflection)
+                .add_service(crate::daemon::proto::turbod_server::TurbodServer::new(self))
+                .serve_with_shutdown("127.0.0.1:5000".parse().unwrap(), shutdown_fut)
+        };
+
+        #[cfg(not(feature = "http"))]
+        let (_lock, server_fut) = {
+            let (lock, stream) = crate::daemon::endpoint::open_socket(self.daemon_root.clone())
+                .await
+                .unwrap();
+
+            (
+                lock,
+                Server::builder()
+                    .add_service(crate::daemon::proto::turbod_server::TurbodServer::new(self))
+                    .serve_with_incoming(stream),
+            )
+        };
+
+        select! {
+            _ = server_fut => CloseReason::ServerClosed,
+            _ = watcher_fut => CloseReason::WatcherClosed,
+            _ = shutdown_fut => CloseReason::Shutdown,
+            _ = timeout_fut => CloseReason::Timeout,
+            _ = ctrl_c() => CloseReason::Interrupt,
+        }
+
+        // here the stop token is dropped, and the pid lock is dropped
+        // causing them to be cleaned up
+    }
+
+    fn bump_timeout(&self) {
+        self.timeout_time.reset(self.timeout);
+    }
+}
+
+#[tonic::async_trait]
+impl<T: Watcher + Send + 'static> proto::turbod_server::Turbod for DaemonServer<T> {
+    async fn hello(
+        &self,
+        request: tonic::Request<proto::HelloRequest>,
+    ) -> Result<tonic::Response<proto::HelloResponse>, tonic::Status> {
+        self.bump_timeout();
+
+        if request.into_inner().version != get_version() {
+            return Err(tonic::Status::unimplemented("version mismatch"));
+        } else {
+            Ok(tonic::Response::new(proto::HelloResponse {}))
+        }
+    }
+
+    async fn shutdown(
+        &self,
+        _request: tonic::Request<proto::ShutdownRequest>,
+    ) -> Result<tonic::Response<proto::ShutdownResponse>, tonic::Status> {
+        self.shutdown.lock().await.take().map(|s| s.send(()));
```

Comment:
Can you confirm that this triggers an asynchronous shutdown? Otherwise there is a deadlock for a graceful shutdown waiting for this shutdown rpc to finish.

Reply:
This function does not yield after the signal is sent so we know the response is sent before the server can close.

I believe that this would still not be an issue if it did though. I will however move the shutdown handler since as it is current requests are just dropped midflight rather than gracefully allowing them to finish.